bioio_base/benchmark.py

```diff
@@ -0,0 +1,148 @@
+"""
+    A simple benchmarking function (`benchmark()`) can be imported from
+    this file to be used by individual readers to then performance test
+    their individual readers
+"""
+import csv
+import datetime
+import multiprocessing
+import pathlib
+import time
+import tracemalloc
+import typing
+
+import psutil
+
+from .reader import Reader
+
+
+def _all_scenes_read(reader: typing.Type[Reader], test_file: pathlib.Path) -> None:
+    """Read all scenes of the file"""
+    image = reader(test_file)
+    for scene in image.scenes:
+        image.set_scene(scene)
+        image.get_image_data()
+
+
+def _all_scenes_delayed_read(
+    reader: typing.Type[Reader], test_file: pathlib.Path
+) -> None:
+    """Read all scenes of the file delayed"""
+    image = reader(test_file)
+    for scene in image.scenes:
+        image.set_scene(scene)
+        image.get_image_dask_data()
+
+
+def _read_ome_metadata(reader: typing.Type[Reader], test_file: pathlib.Path) -> None:
+    """Read the OME metadata of the image"""
+    try:
+        reader(test_file).ome_metadata
+    except Exception:
+        pass
+
+
+def _format_bytes(num: float, suffix: str = "B") -> str:
+    """Formats the bytes given into a human readable format"""
+    for unit in ("", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"):
+        if abs(num) < 1024.0:
+            return f"{num:3.1f}{unit}{suffix}"
+        num /= 1024.0
+    return f"{num:.1f}Yi{suffix}"
+
+
+def _benchmark_function(
+    prefix: str, fn: typing.Callable
+) -> typing.Dict[str, typing.Union[str, float]]:
+    """
+    Gets performance stats for calling the given function.
+    Prefixes the keys of the result by the prefix given.
+    """
+    tracemalloc.start()
+    start_time = time.perf_counter()
+    fn()
+    end_time = time.perf_counter()
+    _current, peak = tracemalloc.get_traced_memory()
+    tracemalloc.stop()
+    time_elapsed = end_time - start_time
```

Comment:
does the tracemalloc impact performance?  I suppose it doesn't matter much, so long as we are consistent with the measurements.

Reply:
I tested with and without and it didn't _seem_ to but yeah I think so long as we do this the same way each time it shouldn't matter too much.